llm_toolkit/prompt_builder.py

```diff
@@ -543,3 +548,116 @@ def build_fixer_prompt(self, benchmark: Benchmark, raw_code: str,
+    """Reads the template for prompts."""
+    with open(template_file) as file:
+      return file.read()
+
+  def build(self,
+            function_signature: str,
+            target_file_type: FileType,
+            example_pair: list[list[str]],
+            project_example_content: Optional[list[list[str]]] = None,
+            project_context_content: Optional[dict] = None) -> prompts.Prompt:
+    """Constructs a prompt using the templates in |self| and saves it."""
+
+    with open(self.priming_template_file, 'r') as f:
+      prompt_text = f.read()
+
+    # Format the priming
+    target_repository = introspector.query_introspector_project_repository(
+        self.benchmark.project)
+    prompt_text = prompt_text.replace('{TARGET_REPO}', target_repository)
+    prompt_text = prompt_text.replace('{TARGET_FUNCTION}',
+                                      self.benchmark.function_signature)
+    function_source = introspector.query_introspector_function_source(
+        self.benchmark.project, self.benchmark.function_signature)
+    prompt_text = prompt_text.replace('{TARGET_FUNCTION_SOURCE_CODE}',
+                                      function_source)
+
+    # Set header avoidance string if there are any headers.
+    headers_to_avoid = introspector.query_introspector_header_files(
+        self.benchmark.project)
+    header_avoid_string = ''
+    if headers_to_avoid:
+      header_avoid_string = ', '.join(headers_to_avoid)
+    prompt_text = prompt_text.replace('{TARGET_HEADER_FILES}',
+                                      header_avoid_string)
+
+    # Add function arg types
+    arg_types = introspector.query_introspector_function_debug_arg_types(
+        self.benchmark.project, self.benchmark.function_signature)
+
+    arg_types_text = ''
+    if arg_types:
+      arg_types_text = 'The target function takes the following arguments:\n'
+      arg_types_text += '- ' + '- '.join(f'{arg}\n' for arg in arg_types)
+
+      arg_types_text += (
+          'You must make sure the arguments passed to the '
+          'function match the types of the function. Do this by casting '
+          'appropriately.\n')
+
+    prompt_text = prompt_text.replace('{FUNCTION_ARG_TYPES_MSG}',
+                                      arg_types_text)
+
+    sample_cross_references = introspector.query_introspector_sample_xrefs(
+        self.benchmark.project, self.benchmark.function_signature)
+    if sample_cross_references:
+      additional_text = (
+          'The target function is used in various places of the target project.'
+          'Please see the following samples of code using the target, which '
+          'you should use as inspiration for the harness to structure the code:'
+          '\n')
+
+      exp_usage = 'Example usage:\n'
+      additional_text += exp_usage + exp_usage.join(
+          f'```c{elem}\n```\n' for elem in sample_cross_references)
+    else:
+      additional_text = ''
```

Comment:
Adding this contextual info (e.g., header files, arg types, refs) is useful, but separating them into a different PR can help us conduct a more controlled experiment to evaluate the effect of c-specific prompts (this PR's main goal).

For example, we first run an experiment with C benchmarks based on your c-specific prompt and compare it against the same benchmarks in our past full experiments.
Then we run another experiment with the contextual info (in a separate PR, for both C and C++ benchmarks) to learn its impact.

Reply:
This is how it was developed though -- by running a lot of experiments on different sets of projects and adjusting based on observations.